internal/app/go-filecoin/connectors/retrieval_market/client.go

```diff
@@ -49,28 +49,32 @@ func NewRetrievalClientConnector(
 }
 
 // GetOrCreatePaymentChannel gets or creates a payment channel and posts to chain
-func (r *RetrievalClientConnector) GetOrCreatePaymentChannel(ctx context.Context, clientAddress address.Address, minerAddress address.Address, clientFundsAvailable abi.TokenAmount, tok shared.TipSetToken) (address.Address, error) {
+func (r *RetrievalClientConnector) GetOrCreatePaymentChannel(ctx context.Context, clientAddress address.Address, minerAddress address.Address, clientFundsAvailable abi.TokenAmount, tok shared.TipSetToken) (address.Address, cid.Cid, error) {
 	if clientAddress == address.Undef || minerAddress == address.Undef {
-		return address.Undef, errors.New("empty address")
+		return address.Undef, cid.Undef, errors.New("empty address")
 	}
 	chinfo, err := r.paychMgr.GetPaymentChannelByAccounts(clientAddress, minerAddress)
 	if err != nil {
-		return address.Undef, err
+		return address.Undef, cid.Undef, err
 	}
 	if chinfo.IsZero() {
 		// create the payment channel
 		bal, err := r.getBalance(ctx, clientAddress, tok)
 		if err != nil {
-			return address.Undef, err
+			return address.Undef, cid.Undef, err
 		}
 
 		filAmt := types.NewAttoFIL(clientFundsAvailable.Int)
 		if bal.LessThan(filAmt) {
-			return address.Undef, errors.New("not enough funds in wallet")
+			return address.Undef, cid.Undef, errors.New("not enough funds in wallet")
 		}
 		return r.paychMgr.CreatePaymentChannel(clientAddress, minerAddress, clientFundsAvailable)
 	}
-	return chinfo.UniqueAddr, nil
+	// TODO: I think this is supposed to return the message CID from the creation of the payment channel.
+	// If there was one already, what CID should this return?
+	// What happens if that message is subsequently re-orged out of the chain and never replayed?
+	// https://github.com/filecoin-project/go-filecoin/issues/4034
```

Comment:
It looks like go-fil-markets ignores the cid if the address is defined. For the second question, our message wait logic could possibly use a confidence interval. It could be implemented with the chain scheduler.

Reply:
I'll leave this comment here with the issue for @shannonwells to confirm if it's ok as is.

I am very uncomfortable with any logic that watches for message CIDs instead of observing the state changes directly. At least unless that logic also handles reverting of that message up to finality in the future. A confidence interval < finality doesn't really help: if/when the interval *is* exceeded, will the module be in a legal state at all?

I'm increasingly aware that the modules use this pattern and am worried.

Reply:
@anorth yes, to be clear, all modules including storage miner wait for messages, and the main change here is that we are no longer have the wait for message hidden inside the CreatePaymentChannel call itself. This allows use to track whether we have at least attempted to post to chain already, and separate the confirmation phase form the posting phase.

We are absolutely working towards the change you describe -- handling reverts up to finality, and handling timeouts in the case of the message never appearing on chain. We have a PR merged in our `development` branch that does much of these this on the storage side. Getting these waiting calls separated from the original posting functions is what this PR does.